Userland/Libraries/LibDisassembly/riscv64/IM.h

```diff
@@ -0,0 +1,305 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "Instruction.h"
+
+// I, M, and Zifencei extensions.
+namespace Disassembly::RISCV64 {
+
+// LUI
+class LoadUpperImmediate : public UJTypeInstruction {
+public:
+    virtual ~LoadUpperImmediate() = default;
+    virtual String to_string(DisplayStyle display_style, u32 origin, Optional<SymbolProvider const&> symbol_provider) const override;
+    virtual String mnemonic() const override;
+    virtual bool instruction_equals(InstructionImpl const&) const override;
+    bool operator==(LoadUpperImmediate const&) const = default;
+
+    LoadUpperImmediate(i32 immediate, Register rd)
+        : UJTypeInstruction(immediate, rd)
+    {
+    }
+};
+
+// JAL
+class JumpAndLink : public UJTypeInstruction {
+public:
+    virtual ~JumpAndLink() = default;
+    virtual String to_string(DisplayStyle display_style, u32 origin, Optional<SymbolProvider const&> symbol_provider) const override;
+    virtual String mnemonic() const override;
+    virtual bool instruction_equals(InstructionImpl const&) const override;
+    bool operator==(JumpAndLink const&) const = default;
+
+    JumpAndLink(i32 immediate, Register rd)
+        : UJTypeInstruction(immediate, rd)
+    {
+    }
+};
+
+// JALR
+class JumpAndLinkRegister : public ITypeInstruction {
+public:
+    virtual ~JumpAndLinkRegister() = default;
+    virtual String to_string(DisplayStyle display_style, u32 origin, Optional<SymbolProvider const&> symbol_provider) const override;
+    virtual String mnemonic() const override;
+    virtual bool instruction_equals(InstructionImpl const&) const override;
+    bool operator==(JumpAndLinkRegister const&) const = default;
+
+    JumpAndLinkRegister(i32 offset, Register base, Register rd)
+        : ITypeInstruction(offset, base, rd)
+    {
+    }
+};
+
+// AUIPC
+class AddUpperImmediateToProgramCounter : public UJTypeInstruction {
+public:
+    virtual ~AddUpperImmediateToProgramCounter() = default;
+    virtual String to_string(DisplayStyle display_style, u32 origin, Optional<SymbolProvider const&> symbol_provider) const override;
+    virtual String mnemonic() const override;
+    virtual bool instruction_equals(InstructionImpl const&) const override;
+    bool operator==(AddUpperImmediateToProgramCounter const&) const = default;
+
+    AddUpperImmediateToProgramCounter(i32 immediate, Register rd)
+        : UJTypeInstruction(immediate, rd)
+    {
+    }
+};
+
+class ArithmeticImmediateInstruction : public ITypeInstruction {
+public:
+    enum class Operation {
+        Add,
+        SetLessThan,
+        SetLessThanUnsigned,
+        Xor,
+        Or,
+        And,
+        ShiftLeftLogical,
+        ShiftRightLogical,
+        ShiftRightArithmetic,
+        AddWord,
+        ShiftLeftLogicalWord,
+        ShiftRightLogicalWord,
+        ShiftRightArithmeticWord,
+    };
+    static bool is_word_operation(Operation);
```

Comment:
This seems unimplemented, even in your full PR.

Reply:
Huh indeed. Maybe that was for formatting purposes originally...